lgc/util/llpcGfxRegHelper.cpp

```diff
@@ -0,0 +1,311 @@
+        break;
+    }
+    return nullptr;
+}
+
+// =====================================================================================================================
+// Set register value
+void SqImgSampRegHelper::SetReg(
+    SqSampRegs id,     // Register ID
+    Value*     pParam) // [in] Param to be set
+{
+    switch (id)
+    {
+    case SqSampRegs::FilterMode:
+    case SqSampRegs::xyMagFilter:
+    case SqSampRegs::xyMinFilter:
+        SetRegCommon(static_cast<uint32_t>(id), pParam);
+        break;
+    default:
+        llvm_unreachable("Set \"IsTileOpt\" is not allowed!");
+        break;
+    }
+}
+
+// =====================================================================================================================
+// SqImgSampReg Bits infomation look up table (Gfx9)
+// Refer to imported/chip/gfx9/gfx9_plus_merged_registers.h : SQ_IMG_RSRC_WORD
+static constexpr BitsInfo g_sqImgRsrcRegBitsGfx9[static_cast<uint32_t>(SqRsrcRegs::SqRsrcRegsCount)] =
+{
+    { 0,  0, 32 }, // BaseAddress
+    { 1,  0,  8 }, // BaseAddressHi
+    { 1, 20,  9 }, // Format
+    { 2,  0, 14 }, // Width
+    { 2, 14, 14 }, // Height
+    { 3,  0, 12 }, // DstSelXYZW
+    { 3, 20,  5 }, // IsTileOpt
+    { 4,  0, 13 }, // Depth
+    { 4, 13, 12 }, // Pitch
+    { 4, 29,  3 }, // BcSwizzle
+    {},            // WidthLo
+    {},            // WidthHi
+};
+
+// =====================================================================================================================
+// SqImgSampReg Bits infomation look up table (Gfx10)
+// Refer to: (Not released)
+static constexpr BitsInfo g_sqImgRsrcRegBitsGfx10[static_cast<uint32_t>(SqRsrcRegs::SqRsrcRegsCount)] =
+{
+    { 0,  0, 32 }, // BaseAddress
+    { 1,  0,  8 }, // BaseAddressHi
+    { 1, 20,  9 }, // Format
+    {},            // Width
+    { 2, 14, 16 }, // Height
+    { 3,  0, 12 }, // DstSelXYZW
+    { 3, 20,  5 }, // IsTileOpt
+    { 4,  0, 16 }, // Depth
+    {},            // Pitch
+    { 3, 25,  3 }, // BcSwizzle
+    { 1, 30,  2 }, // WidthLo
+    { 2,  0, 14 }, // WidthHi
+};
+
+// =====================================================================================================================
+// Helper class for handling Registers defined in SQ_IMG_RSRC_WORD
+SqImgRsrcRegHelper::SqImgRsrcRegHelper(
+    Builder*      pBuilder,      // [in] Bound builder context
+    Value*        pRegister,     // [in] Bound register vec <n x i32>
+    GfxIpVersion* pGfxIpVersion) // [in] Current gfxIp version
+: GfxRegHelper(pBuilder, pRegister)
+{
+    m_pGfxIpVersion = pGfxIpVersion;
+
+    switch (pGfxIpVersion->major)
+    {
+    case 9:
+        m_bitsInfo = g_sqImgRsrcRegBitsGfx9;
+        break;
+    case 10:
+        m_bitsInfo = g_sqImgRsrcRegBitsGfx10;
+        break;
+    default:
+        llvm_unreachable("GfxIp is not supported!");
+        break;
+    }
+    m_bitsState = std::make_unique<BitsState[]>(static_cast<uint32_t>(SqRsrcRegs::SqRsrcRegsCount));
```

Comment:
Actually, I spoke a bit too soon. Is there a way to avoid this dynamic allocation here? Since the size is constant, it should be possible to embed the array directly into the class.

Reply:
Actually I have written a version that use stack rather than new something else. But just considering if the bitsinfo are autogenerated, then maybe the state is too large.
Anyway, it is easy to eliminate the usage of new here, and since the current version contains few bitstates, just be fine.

Reply:
The current version is now without new operation, please review, thanks.